<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>edgenet: regularization with pair-wise interactions • netReg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="edgenet: regularization with pair-wise interactions">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.9.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/edgenet.html">Edgenet</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://www.simon-dirmeier.net">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/dirmeier/netReg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://twitter.com/simon_dirmeier">
    <span class="fa fa fa fa-twitter-square fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>edgenet: regularization with pair-wise interactions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dirmeier/netReg/blob/master/vignettes/edgenet.Rmd"><code>vignettes/edgenet.Rmd</code></a></small>
      <div class="hidden name"><code>edgenet.Rmd</code></div>

    </div>

    
    
<div id="edgenet-tutorial" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#edgenet-tutorial" class="anchor"></a>Edgenet tutorial</h2>
<p>This vignette covers penalization using <code>edgenet</code>. The model is a truncated version from <span class="citation">Cheng et al. (2014)</span>. <code>edgenet</code> uses networks to model pair-wise interactions of covariables and responses in generalized linear regression models. More specifically, <code>edgenet</code> uses a <span class="math inline">\((p \times p)\)</span>-dimensional affinity matrix <span class="math inline">\(\mathbf{G}_X \in \mathbb{R}_+^{p \times p}\)</span> to model the interaction of <span class="math inline">\(p\)</span> covariables in <span class="math inline">\(\mathbf{X} \in \mathbb{R}^{n \times p}\)</span> and analogously a matrix <span class="math inline">\(\mathbf{G}_Y \in \mathbb{R}_+^{q \times q}\)</span> to model interactions of <span class="math inline">\(q\)</span> response variables in <span class="math inline">\(\mathbf{Y} \in \mathbb{R}^{n \times q}\)</span>. The affinity matrices are used for regularization of regression coefficients like this:</p>
<p><span class="math display">\[\begin{equation}
\small
\begin{split}
\hat{\mathbf{B}} = \arg \min_{\mathbf{B}} \,
 &amp; -\ell(\mathbf{B}) + \lambda ||\mathbf{B} ||_1  \\  &amp; +
              \frac{\psi_1}{2}  \sum_{i=1}^p \sum_{j=1}^q
               || \boldsymbol \beta_{i,*} - \boldsymbol \beta_{j,*} ||_2^2 (\mathbf{G}_X)_{i,j}    \\  &amp; +
 \frac{\psi_2}{2}  \sum_{i=1}^p \sum_{j=1}^q || \boldsymbol \beta_{*,i} - \boldsymbol \beta_{*,j} ||_2^2 (\mathbf{G}_Y)_{i,j} \\
 = \arg \min_{\mathbf{B}} \,
 &amp;  -\ell(\mathbf{B})  + \lambda ||\mathbf{B} ||_1  \\  &amp; +
 \psi_1 \text{tr}\left( \mathbf{B}^T \left( \mathbf{D}_{\mathbf{G}_X} - \mathbf{G}_X \right) \mathbf{B}  \right) \\  &amp; +
 \psi_2 \text{tr}\left( \mathbf{B} \left( \mathbf{D}_{\mathbf{G}_Y} - \mathbf{G}_Y  \right) \mathbf{B}^T  \right)
\end{split}
\label{equ:graphreg}
\end{equation}\]</span></p>
<p>Matrix <span class="math inline">\(\mathbf{B} \in \mathbb{R}^{(p + 1) \times q}\)</span> is the matrix of regression coefficients to be estimated (including an intercept). Vectors <span class="math inline">\(\boldsymbol \beta_{i, *}\)</span> and <span class="math inline">\(\boldsymbol \beta_{*, i}\)</span> are the <span class="math inline">\(i\)</span>-th row or column of <span class="math inline">\(\mathbf{B}\)</span>, respectively. Shrinkage <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_1\)</span> and <span class="math inline">\(\psi_2\)</span> are fixed or need to be estimated (e.g., using cross-validation). The log-likelihood function <span class="math inline">\(\ell(\mathbf{B})\)</span> over <span class="math inline">\(q\)</span> exponential family response variables is defined as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ f_{\beta_{*,j}}({y}_{i, j} \mid \mathbf{x}_{i,*})
\end{equation}\]</span></p>
<p>The matrices <span class="math inline">\(\mathbf{D}_{\mathbf{G}} - \mathbf{G}\)</span> are the combinatorial (or normalized) graph Laplacians of an affinity matrix <span class="math inline">\(\mathbf{G}\)</span> <span class="citation">(Chung and Graham 1997)</span>.</p>
<div id="the-gaussian-cause" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#the-gaussian-cause" class="anchor"></a>The Gaussian cause</h3>
<p>In the multivariate Gaussian case, the log-likelihood <span class="math inline">\(\ell\)</span> factorizes as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \mathcal{N}({y}_{i, j} \mid \mathbf{x}_{i,*} \cdot \beta_{*,j}, \sigma_j^2)
\end{equation}\]</span></p>
</div>
<div id="the-binomial-cause" class="section level3 unnumbered">
<h3 class="hasAnchor">
<a href="#the-binomial-cause" class="anchor"></a>The Binomial cause</h3>
<p>In the multivariate Binomial case, the log-likelihood <span class="math inline">\(\ell\)</span> factorizes as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \text{Bernoulli}\left({y}_{i, j} \mid h\left(\mathbf{x}_{i,*} \cdot \beta_{*,j}\right)\right),
\end{equation}\]</span></p>
<p>where <span class="math inline">\(h\)</span> is inverse of a link function, such as the logit or probit.</p>
</div>
<div id="fitting-a-model-to-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fitting-a-model-to-data" class="anchor"></a>Fitting a model to data</h3>
<p>The following section shows how we use <code>edgenet</code>. To fit a model, we first need load the library:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(netReg)</a></code></pre></div>
<p>Set some parameters and create affinity matrices:</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb2-1" data-line-number="1"><span class="co"># parameters</span></a>
<a class="sourceLine" id="cb2-2" data-line-number="2">n &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb2-3" data-line-number="3">p &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb2-4" data-line-number="4">q &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb2-5" data-line-number="5"></a>
<a class="sourceLine" id="cb2-6" data-line-number="6"><span class="co"># affinity matrices</span></a>
<a class="sourceLine" id="cb2-7" data-line-number="7">G.X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/rWishart">rWishart</a></span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(p))[,,<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb2-8" data-line-number="8">G.Y &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/rWishart">rWishart</a></span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(q))[,,<span class="dv">1</span>])</a></code></pre></div>
<p>We created the affinity matrix absolutely random, although in practice a <em>real</em> (biological) observed affinity matrix should be used, because in the end the affinity matrix decides the shrinkage of the coefficients.</p>
<p>The actual fit is straightforward. We create Gaussian data first and then fit the model:</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co"># data</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2">X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>p), n)</a>
<a class="sourceLine" id="cb3-3" data-line-number="3">B &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(p <span class="op">*</span><span class="st"> </span>q), p)</a>
<a class="sourceLine" id="cb3-4" data-line-number="4">Y &lt;-<span class="st"> </span>X <span class="op">%*%</span><span class="st"> </span>B <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>q, <span class="dv">0</span>, <span class="fl">0.1</span>), n)</a>
<a class="sourceLine" id="cb3-5" data-line-number="5"></a>
<a class="sourceLine" id="cb3-6" data-line-number="6">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.X, <span class="dt">G.Y=</span>G.Y, <span class="dt">family=</span>gaussian)</a>
<a class="sourceLine" id="cb3-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      1      1      1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(x) for coefficients</code></pre>
<p>The <code>fit</code> object contains information about coefficients, intercepts etc.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(fit)</a></code></pre></div>
<pre><code>##                     y[1]        y[2]          y[3]        y[4]        y[5]
## (Intercept)  0.008409007  0.00663416 -0.0156761073 -0.01366259 -0.01257649
## x[1]        -1.556517720 -0.07644492 -0.0008311171  0.44344237 -0.18539801
## x[2]        -0.227454275  1.00969827 -0.0872753635  0.36389384  0.20675112
## x[3]         0.219895497 -1.22841132 -0.1896748096  0.33428842  1.52244031
## x[4]        -1.360386372  0.23098685 -0.0991078541 -0.35250175 -2.29486847
## x[5]         0.039294556 -0.67004275  2.0911400318 -0.70881420 -1.01764905
## x[6]        -0.472135872  0.70748007  0.0271250196  0.68965387  1.05857646
## x[7]        -0.571183324 -0.42407316 -1.8531699181 -0.12348060  0.49102539
## x[8]         0.311684698  1.35348511 -0.8344018459  0.23117176  0.16915041
## x[9]        -1.680675387 -1.62058163 -0.4846108854  0.74677157 -0.78819597
## x[10]       -0.400453150 -0.15452631 -0.2470075786 -0.94371283 -0.72495937
##                     y[6]        y[7]         y[8]        y[9]        y[10]
## (Intercept)  0.004503852  0.00409580 -0.003266989  0.01986224 -0.006966052
## x[1]         1.262800217 -0.74232435  0.364759743  0.84148681 -0.623343825
## x[2]         0.197979286  0.07930420  1.057213783 -0.34472618  2.243035078
## x[3]         0.138299301  0.02495443  1.075191736  0.78341579 -0.447208703
## x[4]        -0.022121629  0.43837619  1.186983466  1.02085769 -0.967668295
## x[5]         0.028485943  1.27903485 -0.206207812 -0.22584569  0.588629246
## x[6]        -0.261091948 -0.04609604 -0.243982449  1.13111317 -0.878119767
## x[7]         0.619860470  1.54257286 -0.247289076 -2.41254163  0.360397518
## x[8]         1.588969707  0.23127513 -1.240848422  1.75545204  0.744822025
## x[9]         0.362172037  1.09318447  1.144388318 -0.19231145 -0.061439134
## x[10]       -0.541737378  2.06742644  0.305232733 -1.03793883 -1.018907905</code></pre>
<p>Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb7-1" data-line-number="1">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(fit, X)</a>
<a class="sourceLine" id="cb7-2" data-line-number="2">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##            [,1]       [,2]      [,3]       [,4]       [,5]
## [1,]  0.3377337 -0.8307156 -1.895531  0.6536721  0.2501273
## [2,]  3.8466857  2.3340276  6.564003  0.5349036  1.5641865
## [3,]  2.9862713  6.2292793  1.919214 -0.0838942 -1.3075693
## [4,] -1.1452688 -0.3421425  4.067183 -0.6829166 -3.6184361
## [5,]  1.9623935 -2.0184022  3.555467 -3.0846856 -3.9806147</code></pre>
<p>The binomial case is the same only with a different <code>family</code> argument:</p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb9-1" data-line-number="1"><span class="co"># data</span></a>
<a class="sourceLine" id="cb9-2" data-line-number="2">X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>p), n)</a>
<a class="sourceLine" id="cb9-3" data-line-number="3">B &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(p <span class="op">*</span><span class="st"> </span>q), p)</a>
<a class="sourceLine" id="cb9-4" data-line-number="4"></a>
<a class="sourceLine" id="cb9-5" data-line-number="5">eta &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="op">-</span>X <span class="op">%*%</span><span class="st"> </span>B))</a>
<a class="sourceLine" id="cb9-6" data-line-number="6">Y.binom &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="st">"cbind"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">10</span>), <span class="cf">function</span>(.) <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(n, <span class="dv">1</span>, eta[,.])))</a>
<a class="sourceLine" id="cb9-7" data-line-number="7"></a>
<a class="sourceLine" id="cb9-8" data-line-number="8">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.X, <span class="dt">G.Y=</span>G.Y, <span class="dt">family=</span>binomial)</a>
<a class="sourceLine" id="cb9-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'binomial.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, family = binomial)
## 
## parameters:
## lambda  psigx  psigy 
##      1      1      1 
## 
## family:  binomial
## link:  logit 
## 
## -&gt; call coef(x) for coefficients</code></pre>
</div>
<div id="model-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#model-selection" class="anchor"></a>Model selection</h3>
<p>In most cases we do not have the optimal shrinkage parameters <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_{1}\)</span> and <span class="math inline">\(\psi_{2}\)</span>. For these settings you can use <code>netReg</code>’s included model selection. We use Powell’s BOBYQA algorithm <span class="citation">(Powell 2009)</span> for gradient-free optimization in a coss-validation framework. Doing the model selection only requires calling <code>cv.edgenet</code>:</p>
<div class="sourceCode" id="cb11"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb11-1" data-line-number="1">cv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cv.edgenet.html">cv.edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.Y, G.Y, <span class="dt">family=</span>gaussian, <span class="dt">optim.maxit=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb11-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(cv)</a></code></pre></div>
<pre><code>## 'gaussian.cv.edgenet' object
## 
## call:
## cv.edgenet(X = X, Y = Y, G.X = G.Y, G.Y = G.Y, family = gaussian, 
##     optim.maxit = 10)
## 
## optimal parameters:
## lambda (estimated)  psigx (estimated)  psigy (estimated) 
##         0.00000000         0.02928932         0.10000000 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(x) for coefficients</code></pre>
<p>The cross-validation and BOBYQA is quite computationally intensive, so we set <code>optim.maxit</code> to <span class="math inline">\(10\)</span>. I recommend using <code>TensorFlow</code> on a GPU for that, since <code>edgenet</code> needs to compute many matrix multiplications.</p>
<p>The <code>cv.edgenet</code> object also contains a fit with the optimal parameters.</p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(cv<span class="op">$</span>fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, lambda = ret$lambda, 
##     psigx = ret$psigx, psigy = ret$psigy, thresh = thresh, maxit = maxit, 
##     learning.rate = learning.rate, family = family)
## 
## parameters:
##     lambda      psigx      psigy 
## 0.00000000 0.02928932 0.10000000 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(x) for coefficients</code></pre>
<p>You can obtain the coefficients like this:</p>
<div class="sourceCode" id="cb15"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb15-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(cv)</a></code></pre></div>
<pre><code>##                    y[1]        y[2]        y[3]        y[4]         y[5]
## (Intercept) -0.15814951  0.27650547 -0.40680209 -0.11027102 -0.201240316
## x[1]        -0.06985629 -0.05583192  0.13417025 -0.04171887 -0.111750528
## x[2]        -0.28642604  0.26652259 -0.29621217 -0.02490871 -0.253131002
## x[3]         0.21639602  0.06991196  0.01115142 -0.02145038  0.008819572
## x[4]         0.39742571  0.16338693 -0.09901705  0.12263419  0.321040004
## x[5]        -0.26956084  0.35882047 -0.11949764 -0.08983727 -0.476000130
## x[6]         0.37196678 -0.16404492  0.64811301 -0.18785189 -0.061399944
## x[7]        -0.15813529 -0.09122412  0.26971272 -0.06773197  0.077435762
## x[8]        -0.61610174  0.01122050 -0.54709154 -0.06667845 -0.665743113
## x[9]        -0.14399502 -0.36602733  0.13350751 -0.11386552  0.045257706
## x[10]        0.27761650 -0.26351941  0.16711125 -0.26952496  0.119324282
##                    y[6]        y[7]          y[8]        y[9]       y[10]
## (Intercept) -0.03126104 -0.03623462  0.2176410854 -0.20338151 -0.08253793
## x[1]         0.06469074  0.02195547 -0.1112701073  0.02415625 -0.08389207
## x[2]         0.36537737 -0.09567580  0.0009108149  0.43088394 -0.24790911
## x[3]        -0.08816558  0.03414125  0.0777896047  0.59039348 -0.13393436
## x[4]        -0.21393858 -0.26533365  0.2279571295  0.40373743  0.25088409
## x[5]         0.02056166  0.17800419 -0.0114902155  0.45586786 -0.35671151
## x[6]        -0.40121126 -0.36237234 -0.0695453882 -0.07654646 -0.11587922
## x[7]        -0.04768996  0.41528839  0.1787957400 -0.40413246  0.21183011
## x[8]         0.45999390  0.42561239  0.0723909438 -0.26599911  0.04568073
## x[9]        -0.04955092  0.33121371 -0.1465795189 -0.66557831 -0.16131926
## x[10]       -0.19812609  0.40474886 -0.0711703077 -0.65543652  0.15971655</code></pre>
<p>Furthermore, you can also directly predict using a <code>cv.edgenet</code> object:</p>
<div class="sourceCode" id="cb17"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb17-1" data-line-number="1">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(cv, X)</a>
<a class="sourceLine" id="cb17-2" data-line-number="2">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##            [,1]       [,2]        [,3]       [,4]       [,5]
## [1,] -2.1935470 0.05059089 -1.35901550 -0.1180447 -1.3592006
## [2,] -0.1247577 0.02397266  0.04344704 -0.3423867  0.1676264
## [3,] -0.3320200 0.63517767 -1.32792809  0.4923493  0.1621701
## [4,]  0.2748583 0.18221666 -1.01030419  0.2535366  0.3194843
## [5,]  1.5836006 0.25803539  0.19289459  0.1504275  1.5496119</code></pre>
</div>
<div id="a-biological-example" class="section level3">
<h3 class="hasAnchor">
<a href="#a-biological-example" class="anchor"></a>A biological example</h3>
<p>This section explains how to fit a linear model and do parameter estimation.</p>
<p>At first we load the library and some data:</p>
<div class="sourceCode" id="cb19"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb19-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(netReg)</a>
<a class="sourceLine" id="cb19-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"yeast"</span>)</a>
<a class="sourceLine" id="cb19-3" data-line-number="3"></a>
<a class="sourceLine" id="cb19-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>(yeast)</a>
<a class="sourceLine" id="cb19-5" data-line-number="5"></a>
<a class="sourceLine" id="cb19-6" data-line-number="6">X &lt;-<span class="st"> </span>yeast<span class="op">$</span>X</a>
<a class="sourceLine" id="cb19-7" data-line-number="7">Y &lt;-<span class="st"> </span>yeast<span class="op">$</span>Y</a>
<a class="sourceLine" id="cb19-8" data-line-number="8">G.Y &lt;-<span class="st"> </span>yeast<span class="op">$</span>GY</a></code></pre></div>
<p>The <em>yeast</em> data <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> set is taken and adopted from <span class="citation">(Brem et al. 2005)</span>, <span class="citation">(Storey, Akey, and Kruglyak 2005)</span>, and <span class="citation">(Cheng et al. 2014)</span>. <span class="math inline">\(\mathbf{GY}\)</span> is taken from <a href="https://thebiogrid.org/downloads/archives/Release%20Archive/BIOGRID-3.4.150/BIOGRID-ORGANISM-3.4.150.tab.zip">BioGRID</a>.</p>
<p><span class="math inline">\(\mathbf{X}\)</span> is a <span class="math inline">\((n \times p)\)</span> matrix of genetic markers where <span class="math inline">\(n\)</span> is the number of samples (112) and <span class="math inline">\(p\)</span> is the number of markers. <span class="math inline">\(\mathbf{Y}\)</span> is a <span class="math inline">\((n \times q)\)</span> matrix of expression values for <span class="math inline">\(q\)</span> yeast genes. <span class="math inline">\(n\)</span> is again the numer of samples (112). <span class="math inline">\(\mathbf{GY}\)</span> is a <span class="math inline">\((q \times q)\)</span> adjacency matrix representing protein-protein interactions for the <span class="math inline">\(q\)</span> response variables.</p>
<p>We only use a smaller network in order to be able to print the results here.</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.Y=</span>G.Y, <span class="dt">lambda=</span><span class="dv">5</span>, <span class="dt">family=</span>gaussian, <span class="dt">maxit=</span><span class="dv">100</span>, <span class="dt">thresh=</span><span class="fl">1e-3</span>)</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.Y = G.Y, lambda = 5, thresh = 0.001, 
##     maxit = 100, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      5      0      1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(x) for coefficients</code></pre>
<p>For the response variables we use an affinity matrix that represents <em>biological relationships</em> with <span class="math inline">\(\mathbf{GY}\)</span>. We promote sparsity by setting <span class="math inline">\(\lambda = 5\)</span> and put a weak (default) penalty on similar coefficients by setting <span class="math inline">\(\psi_{gy} = 1\)</span>. Other than that we used standard parameters in this case.</p>
<p>The <code>fit</code> object contains information about coefficients and intercepts. Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">X.new &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(<span class="dv">10</span> <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(X)), <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(fit, X.new)</a>
<a class="sourceLine" id="cb22-3" data-line-number="3">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]</a></code></pre></div>
<pre><code>##             [,1]        [,2]         [,3]        [,4]       [,5]
##  [1,] -0.1449338  0.88943465  0.364707446 -0.12325593 -0.3658354
##  [2,]  0.1948131  0.79343326  0.002480588  0.26535934  0.4670230
##  [3,] -0.8056121  0.36252469 -0.220239089 -0.47536074 -0.8260944
##  [4,] -1.5536845  0.88793776 -1.355180019 -0.10949346 -1.8130933
##  [5,] -0.4452423  0.03728228  0.182626325 -0.38252106 -0.2996882
##  [6,] -0.3716727 -0.37933740 -0.113981341 -0.43195586 -0.3492265
##  [7,]  0.1428319  0.01772786  0.158745067 -0.34579042 -0.2135023
##  [8,]  0.5414015  0.33700398  0.756357282  0.08741236  0.2454708
##  [9,]  1.1838071  0.31640016  0.531059663 -0.43407075  0.4702082
## [10,] -0.3544736  0.49112456  0.388560179 -0.23892109 -0.3065714
##             [,6]        [,7]        [,8]       [,9]       [,10]
##  [1,] -0.6356759  0.29735774  0.41412174  0.8879513 -0.57710063
##  [2,] -0.3827783  0.02989838  0.75750135  0.7299494  0.57625428
##  [3,]  0.2176652  0.65181469  0.59430334 -1.2730587  0.23035265
##  [4,]  0.7613639  0.28098664  0.65883967  0.5893891 -0.49482531
##  [5,] -0.1178086  0.37475968  0.15977543 -1.6195011  0.11868061
##  [6,]  0.1316399  0.66205316  0.02820699 -1.2734278 -0.12625701
##  [7,] -0.3335304 -0.02001579  0.10690927 -0.2350669 -0.36492380
##  [8,] -0.5611707 -1.73103899 -0.05569267  0.4786942 -0.10637181
##  [9,] -1.0435450 -1.01543946  0.21211574 -0.1925221  0.04157173
## [10,] -0.2075858  0.20477328  0.38287913 -0.1627889 -0.23519000</code></pre>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-brem2005genetic">
<p>Brem, Rachel B, John D Storey, Jacqueline Whittle, and Leonid Kruglyak. 2005. “Genetic Interactions Between Polymorphisms That Affect Gene Expression in Yeast.” <em>Nature</em> 436 (7051). NIH Public Access: 701.</p>
</div>
<div id="ref-cheng2014graph">
<p>Cheng, Wei, Xiang Zhang, Zhishan Guo, Yu Shi, and Wei Wang. 2014. “Graph-Regularized Dual Lasso for Robust eQTL Mapping.” <em>Bioinformatics</em> 30 (12). Oxford Univ Press: i139–i148.</p>
</div>
<div id="ref-chung1997spectral">
<p>Chung, Fan RK, and Fan Chung Graham. 1997. <em>Spectral Graph Theory</em>. 92. American Mathematical Soc.</p>
</div>
<div id="ref-powell2009bobyqa">
<p>Powell, Michael JD. 2009. “The Bobyqa Algorithm for Bound Constrained Optimization Without Derivatives.” <em>Cambridge NA Report NA2009/06, University of Cambridge, Cambridge</em>. Citeseer.</p>
</div>
<div id="ref-storey2005multiple">
<p>Storey, John D, Joshua M Akey, and Leonid Kruglyak. 2005. “Multiple Locus Linkage Analysis of Genomewide Expression in Yeast.” <em>PLoS Biology</em> 3 (8). Public Library of Science: e267.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#edgenet-tutorial">Edgenet tutorial</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Simon Dirmeier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
