<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>edgenet: regularization with pair-wise interactions • netReg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.3.1/jquery.min.js" integrity="sha256-FgpCb/KJQlLNfOu91ta32o/NMZxltwRo8QtmkMRdAu8=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.3.7/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha256-U5ZEeKfGNOja007MMD3YBI0A3OSZOQbeG6z2f2Y0hu8=" crossorigin="anonymous"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/4.7.0/css/font-awesome.min.css" integrity="sha256-eZrrJcwDc/3uDhsdt61sL2oOBY362qM3lon1gyExkL0=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.4/clipboard.min.js" integrity="sha256-FiZwavyI2V6+EXO1U+xzLG3IKldpiTFf3153ea9zikQ=" crossorigin="anonymous"></script><!-- sticky kit --><script src="https://cdnjs.cloudflare.com/ajax/libs/sticky-kit/1.1.3/sticky-kit.min.js" integrity="sha256-c4Rlo1ZozqTPE2RLuvbusY3+SU1pQaJC0TjuhygMipw=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="edgenet: regularization with pair-wise interactions">
<meta property="og:description" content="">
<meta name="twitter:card" content="summary">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.11.1</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/edgenet.html">Edgenet</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://www.simon-dirmeier.net">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/dirmeier/netReg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://twitter.com/simon_dirmeier">
    <span class="fa fa fa fa-twitter-square fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1>edgenet: regularization with pair-wise interactions</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dirmeier/netReg/blob/master/vignettes/edgenet.Rmd"><code>vignettes/edgenet.Rmd</code></a></small>
      <div class="hidden name"><code>edgenet.Rmd</code></div>

    </div>

    
    
<div id="edgenet-tutorial" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#edgenet-tutorial" class="anchor"></a>Edgenet tutorial</h2>
<p>This vignette covers penalization using <code>edgenet</code> as introduced in <span class="citation">Dirmeier et al. (2018)</span>. <code>edgenet</code> uses networks to model pair-wise interactions of covariables and responses in generalized linear regression models. More specifically, <code>edgenet</code> uses a <span class="math inline">\((p \times p)\)</span>-dimensional affinity matrix <span class="math inline">\(\mathbf{G}_X \in \mathbb{R}_+^{p \times p}\)</span> to model the interaction of <span class="math inline">\(p\)</span> covariables in <span class="math inline">\(\mathbf{X} \in \mathbb{R}^{n \times p}\)</span> and analogously a matrix <span class="math inline">\(\mathbf{G}_Y \in \mathbb{R}_+^{q \times q}\)</span> to model interactions of <span class="math inline">\(q\)</span> response variables in <span class="math inline">\(\mathbf{Y} \in \mathbb{R}^{n \times q}\)</span>. The affinity matrices are used for regularization of regression coefficients like this:</p>
<p><span class="math display">\[\begin{equation}
\small
\begin{split}
\hat{\mathbf{B}} = \arg \min_{\mathbf{B}} \,
 &amp; -\ell(\mathbf{B}) + \lambda ||\mathbf{B} ||_1  \\  &amp; +
              \frac{\psi_1}{2}  \sum_{i=1}^p \sum_{j=1}^p
               || \boldsymbol \beta_{i,*} - \boldsymbol \beta_{j,*} ||_2^2 (\mathbf{G}_X)_{i,j}    \\  &amp; +
 \frac{\psi_2}{2}  \sum_{i=1}^q \sum_{j=1}^q || \boldsymbol \beta_{*,i} - \boldsymbol \beta_{*,j} ||_2^2 (\mathbf{G}_Y)_{i,j} \\
 = \arg \min_{\mathbf{B}} \,
 &amp;  -\ell(\mathbf{B})  + \lambda ||\mathbf{B} ||_1  \\  &amp; +
 \psi_1 \text{tr}\left( \mathbf{B}^T \left( \mathbf{D}_{\mathbf{G}_X} - \mathbf{G}_X \right) \mathbf{B}  \right) \\  &amp; +
 \psi_2 \text{tr}\left( \mathbf{B} \left( \mathbf{D}_{\mathbf{G}_Y} - \mathbf{G}_Y  \right) \mathbf{B}^T  \right)
\end{split}
\label{equ:graphreg}
\end{equation}\]</span></p>
<p>Matrix <span class="math inline">\(\mathbf{B} \in \mathbb{R}^{(p + 1) \times q}\)</span> is the matrix of regression coefficients to be estimated (including an intercept). Vectors <span class="math inline">\(\boldsymbol \beta_{i, *}\)</span> and <span class="math inline">\(\boldsymbol \beta_{*, i}\)</span> are the <span class="math inline">\(i\)</span>-th row or column of <span class="math inline">\(\mathbf{B}\)</span>, respectively. Shrinkage <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_1\)</span> and <span class="math inline">\(\psi_2\)</span> are fixed or need to be estimated (e.g., using cross-validation). The log-likelihood function <span class="math inline">\(\ell(\mathbf{B})\)</span> over <span class="math inline">\(q\)</span> exponential family response variables is defined as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ f_{\beta_{*,j}}({y}_{i, j} \mid \mathbf{x}_{i,*})
\end{equation}\]</span></p>
<p>The matrices <span class="math inline">\(\mathbf{D}_{\mathbf{G}} - \mathbf{G}\)</span> are the combinatorial (or normalized) graph Laplacians of an affinity matrix <span class="math inline">\(\mathbf{G}\)</span> <span class="citation">(Chung and Graham 1997)</span>.</p>
<div id="families" class="section level3">
<h3 class="hasAnchor">
<a href="#families" class="anchor"></a>Families</h3>
<p>So far <code>edgenet</code> supports the following families:</p>
<ul>
<li>Gaussian,</li>
<li>Binomial,</li>
<li>Poisson.</li>
</ul>
<div id="the-gaussian-case" class="section level5 unnumbered">
<h5 class="hasAnchor">
<a href="#the-gaussian-case" class="anchor"></a>The Gaussian case</h5>
<p>In the multivariate Gaussian case, the log-likelihood <span class="math inline">\(\ell\)</span> factorizes as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \mathcal{N}({y}_{i, j} \mid \mathbf{x}_{i,*} \cdot \beta_{*,j}, \sigma_j^2)
\end{equation}\]</span></p>
</div>
<div id="the-binomial-case" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#the-binomial-case" class="anchor"></a>The Binomial case</h4>
<p>In the multivariate Binomial case, the log-likelihood <span class="math inline">\(\ell\)</span> factorizes as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \text{Bernoulli}\left({y}_{i, j} \mid h\left(\mathbf{x}_{i,*} \cdot \beta_{*,j}\right)\right),
\end{equation}\]</span></p>
<p>where <span class="math inline">\(h\)</span> is the inverse of a link function, such as the logit or probit.</p>
</div>
<div id="the-poisson-case" class="section level4 unnumbered">
<h4 class="hasAnchor">
<a href="#the-poisson-case" class="anchor"></a>The Poisson case</h4>
<p>In the multivariate Poisson case, the log-likelihood <span class="math inline">\(\ell\)</span> factorizes as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \text{Pois}\left({y}_{i, j} \mid h\left(\mathbf{x}_{i,*} \cdot \beta_{*,j}\right)\right),
\end{equation}\]</span></p>
<p>where <span class="math inline">\(h\)</span> is the inverse of a link function, such as the logarithm.</p>
</div>
</div>
<div id="fitting-a-model-to-data" class="section level3">
<h3 class="hasAnchor">
<a href="#fitting-a-model-to-data" class="anchor"></a>Fitting a model to data</h3>
<p>The following section shows how we use <code>edgenet</code>. To fit a model, we first need load the library:</p>
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb1-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(netReg)</a></code></pre></div>
<pre><code>## 
## Attaching package: 'netReg'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     binomial, family, gaussian, inverse.gaussian, poisson</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     beta, gamma</code></pre>
<p>Set some parameters and create affinity matrices:</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># parameters</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">n &lt;-<span class="st"> </span><span class="dv">100</span></a>
<a class="sourceLine" id="cb5-3" data-line-number="3">p &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb5-4" data-line-number="4">q &lt;-<span class="st"> </span><span class="dv">10</span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"></a>
<a class="sourceLine" id="cb5-6" data-line-number="6"><span class="co"># affinity matrices</span></a>
<a class="sourceLine" id="cb5-7" data-line-number="7">G.X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/rWishart">rWishart</a></span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(p))[,,<span class="dv">1</span>])</a>
<a class="sourceLine" id="cb5-8" data-line-number="8">G.Y &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/MathFun">abs</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/rWishart">rWishart</a></span>(<span class="dv">1</span>, <span class="dv">10</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/diag">diag</a></span>(q))[,,<span class="dv">1</span>])</a></code></pre></div>
<p>We created the affinity matrix absolutely random, although in practice a <em>real</em> (biological) observed affinity matrix should be used, because in the end the affinity matrix decides the shrinkage of the coefficients.</p>
<p>The actual fit is straightforward. We create Gaussian data first and then fit the model:</p>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="co"># data</span></a>
<a class="sourceLine" id="cb6-2" data-line-number="2">X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>p), n)</a>
<a class="sourceLine" id="cb6-3" data-line-number="3">B &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(p <span class="op">*</span><span class="st"> </span>q), p)</a>
<a class="sourceLine" id="cb6-4" data-line-number="4">Y &lt;-<span class="st"> </span>X <span class="op">%*%</span><span class="st"> </span>B <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>q, <span class="dv">0</span>, <span class="fl">0.1</span>), n)</a>
<a class="sourceLine" id="cb6-5" data-line-number="5"></a>
<a class="sourceLine" id="cb6-6" data-line-number="6">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.X, <span class="dt">G.Y=</span>G.Y, <span class="dt">family=</span>gaussian, <span class="dt">maxit=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb6-7" data-line-number="7"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, maxit = 10, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      1      1      1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>The <code>fit</code> object contains information about coefficients, intercepts etc.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(fit)[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##                  y[1]      y[2]      y[3]      y[4]      y[5]
## (Intercept) 0.9003567 0.9003829 0.9003120 0.9002066 0.9003463
## x[1]        0.9002273 0.9002372 1.0996169 0.9001074 0.9004163
## x[2]        0.9006475 0.9010343 1.0994906 0.9022833 0.9007675
## x[3]        1.0996490 0.9002360 0.9002746 0.9001336 0.9006982
## x[4]        0.9001578 0.9002684 0.9063749 0.9021673 0.9001877
## x[5]        0.9056488 1.0991423 0.9003448 0.9006447 0.9007891
## x[6]        0.9012993 0.9019507 0.9001886 0.9007923 0.9002187
## x[7]        0.9007698 0.9009798 1.0982603 1.0988039 1.0982139
## x[8]        0.9383100 0.9004335 0.9003744 0.9006745 0.9001517
## x[9]        0.9005579 0.9003015 0.9002422 0.9004556 0.9003394
## x[10]       1.0996001 1.0995895 0.9002583 1.0996617 0.9008908</code></pre>
<p>Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(fit, X)</a>
<a class="sourceLine" id="cb10-2" data-line-number="2">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##            [,1]      [,2]       [,3]        [,4]       [,5]
## [1,]  1.4492133  1.329955  0.6271177  1.07578802 1.06648163
## [2,]  0.9663546  1.499443  1.0397122  1.24714198 1.15415416
## [3,]  1.9908291  1.935616  2.5117548  2.37106225 2.56071533
## [4,]  5.2396438  5.111284  6.2007282  5.11137390 5.44285744
## [5,] -0.1095857 -0.415109 -0.2209881 -0.07265518 0.06788154</code></pre>
<p>The binomial case is the same only with a different <code>family</code> argument:</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># data</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2">X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>p), n)</a>
<a class="sourceLine" id="cb12-3" data-line-number="3">B &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(p <span class="op">*</span><span class="st"> </span>q), p)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5">eta &lt;-<span class="st"> </span><span class="dv">1</span> <span class="op">/</span><span class="st"> </span>(<span class="dv">1</span> <span class="op">+</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="op">-</span>X <span class="op">%*%</span><span class="st"> </span>B))</a>
<a class="sourceLine" id="cb12-6" data-line-number="6">Y.binom &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="st">"cbind"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">10</span>), <span class="cf">function</span>(.) <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Binomial">rbinom</a></span>(n, <span class="dv">1</span>, eta[,.])))</a>
<a class="sourceLine" id="cb12-7" data-line-number="7"></a>
<a class="sourceLine" id="cb12-8" data-line-number="8">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.X, <span class="dt">G.Y=</span>G.Y, <span class="dt">family=</span>binomial, <span class="dt">maxit=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb12-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'binomial.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, maxit = 10, family = binomial)
## 
## parameters:
## lambda  psigx  psigy 
##      1      1      1 
## 
## family:  binomial
## link:  logit 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>The Poisson case of course works analogously:</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="co"># data</span></a>
<a class="sourceLine" id="cb14-2" data-line-number="2">X &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(n <span class="op">*</span><span class="st"> </span>p), n)</a>
<a class="sourceLine" id="cb14-3" data-line-number="3">B &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(p <span class="op">*</span><span class="st"> </span>q), p)</a>
<a class="sourceLine" id="cb14-4" data-line-number="4"></a>
<a class="sourceLine" id="cb14-5" data-line-number="5">eta &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/Log">exp</a></span>(<span class="op">-</span>X <span class="op">%*%</span><span class="st"> </span>B)</a>
<a class="sourceLine" id="cb14-6" data-line-number="6">Y.pois &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/do.call">do.call</a></span>(<span class="st">"cbind"</span>, <span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/lapply">lapply</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/seq">seq</a></span>(<span class="dv">10</span>), <span class="cf">function</span>(.) <span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Poisson">rpois</a></span>(n, eta[,.])))</a>
<a class="sourceLine" id="cb14-7" data-line-number="7"></a>
<a class="sourceLine" id="cb14-8" data-line-number="8">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y.pois, <span class="dt">G.X=</span>G.X, <span class="dt">G.Y=</span>G.Y, <span class="dt">family=</span>poisson, <span class="dt">maxit=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb14-9" data-line-number="9"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'poisson.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y.pois, G.X = G.X, G.Y = G.Y, maxit = 10, 
##     family = poisson)
## 
## parameters:
## lambda  psigx  psigy 
##      1      1      1 
## 
## family:  poisson
## link:  log 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
</div>
<div id="model-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#model-selection" class="anchor"></a>Model selection</h3>
<p>In most cases we do not have the optimal shrinkage parameters <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_{1}\)</span> and <span class="math inline">\(\psi_{2}\)</span>. For these settings you can use <code>netReg</code>’s included model selection. We use Powell’s BOBYQA algorithm <span class="citation">(Powell 2009)</span> for gradient-free optimization in a coss-validation framework. Doing the model selection only requires calling <code>cv.edgenet</code>:</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1">cv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cvedgenet-methods.html">cv.edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>G.Y, G.Y, <span class="dt">family=</span>gaussian, <span class="dt">optim.maxit=</span><span class="dv">10</span>, <span class="dt">maxit=</span><span class="dv">10</span>)</a>
<a class="sourceLine" id="cb16-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(cv)</a></code></pre></div>
<pre><code>## 'gaussian.cv.edgenet' object
## 
## call:
## cv.edgenet(X = X, Y = Y, G.X = G.Y, G.Y = G.Y, maxit = 10, family = gaussian, 
##     optim.maxit = 10)
## 
## optimal parameters:
## lambda (estimated)  psigx (estimated)  psigy (estimated) 
##         0.00000000         0.02928932         0.10000000 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(cv) for coefficients</code></pre>
<p>The cross-validation and BOBYQA is quite computationally intensive, so we set <code>optim.maxit</code> to <span class="math inline">\(10\)</span>. I recommend using <code>TensorFlow</code> on a GPU for that, since <code>edgenet</code> needs to compute many matrix multiplications.</p>
<p>The <code>cv.edgenet</code> object also contains a fit with the optimal parameters.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(cv<span class="op">$</span>fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, lambda = ret$lambda, 
##     psigx = ret$psigx, psigy = ret$psigy, thresh = thresh, maxit = maxit, 
##     learning.rate = learning.rate, family = family)
## 
## parameters:
##     lambda      psigx      psigy 
## 0.00000000 0.02928932 0.10000000 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(cv$fit) for coefficients</code></pre>
<p>You can obtain the coefficients like this:</p>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/coef">coef</a></span>(cv)[,<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##                  y[1]      y[2]      y[3]      y[4]      y[5]
## (Intercept) 0.9003022 0.9003109 0.9003357 0.9002512 0.9003797
## x[1]        0.9003770 0.9004253 0.9003255 0.9004640 0.9003642
## x[2]        0.9003659 0.9003244 0.9004520 0.9004127 0.9003683
## x[3]        0.9005399 0.9004351 0.9002484 0.9004454 0.9003146
## x[4]        0.9002989 0.9003041 0.9004574 0.9004608 0.9003270
## x[5]        0.9003553 0.9003905 0.9004829 0.9003624 0.9004826
## x[6]        0.9016979 0.9009241 0.9002509 0.9006585 0.9003376
## x[7]        0.9005997 0.9009412 0.9002198 0.9017729 0.9005366
## x[8]        0.9005347 0.9003763 0.9003918 0.9003374 0.9004546
## x[9]        0.9007616 0.9003509 0.9003669 0.9002401 0.9003268
## x[10]       0.9006305 0.9017265 0.9001973 0.9019073 0.9002485</code></pre>
<p>Furthermore, you can also directly predict using a <code>cv.edgenet</code> object:</p>
<div class="sourceCode" id="cb22"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb22-1" data-line-number="1">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(cv, X)</a>
<a class="sourceLine" id="cb22-2" data-line-number="2">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">5</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##             [,1]        [,2]        [,3]        [,4]        [,5]
## [1,]  2.61235412  2.61375459  2.61228128  2.61446516  2.61254287
## [2,] -0.07524288 -0.07389706 -0.07414461 -0.07459006 -0.07442647
## [3,]  1.39466527  1.39543944  1.39464965  1.39418459  1.39441324
## [4,]  1.94471879  1.94350577  1.94451071  1.94445160  1.94510785
## [5,]  5.64740243  5.64910908  5.64743100  5.65057062  5.64806896</code></pre>
</div>
<div id="a-biological-example" class="section level3">
<h3 class="hasAnchor">
<a href="#a-biological-example" class="anchor"></a>A biological example</h3>
<p>This section explains how to fit a linear model and do parameter estimation.</p>
<p>At first we load the library and some data:</p>
<div class="sourceCode" id="cb24"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb24-1" data-line-number="1"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/library">library</a></span>(netReg)</a>
<a class="sourceLine" id="cb24-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/utils/topics/data">data</a></span>(<span class="st">"yeast"</span>)</a>
<a class="sourceLine" id="cb24-3" data-line-number="3"></a>
<a class="sourceLine" id="cb24-4" data-line-number="4"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/ls">ls</a></span>(yeast)</a>
<a class="sourceLine" id="cb24-5" data-line-number="5"></a>
<a class="sourceLine" id="cb24-6" data-line-number="6">X &lt;-<span class="st"> </span>yeast<span class="op">$</span>X</a>
<a class="sourceLine" id="cb24-7" data-line-number="7">Y &lt;-<span class="st"> </span>yeast<span class="op">$</span>Y</a>
<a class="sourceLine" id="cb24-8" data-line-number="8">G.Y &lt;-<span class="st"> </span>yeast<span class="op">$</span>GY</a></code></pre></div>
<p>The <em>yeast</em> data <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> set is taken and adopted from <span class="citation">(Brem et al. 2005)</span>, <span class="citation">(Storey, Akey, and Kruglyak 2005)</span>, and <span class="citation">(Cheng et al. 2014)</span>. <span class="math inline">\(\mathbf{GY}\)</span> is taken from <a href="https://thebiogrid.org/downloads/archives/Release%20Archive/BIOGRID-3.4.150/BIOGRID-ORGANISM-3.4.150.tab.zip">BioGRID</a>.</p>
<p><span class="math inline">\(\mathbf{X}\)</span> is a <span class="math inline">\((n \times p)\)</span> matrix of genetic markers where <span class="math inline">\(n\)</span> is the number of samples (112) and <span class="math inline">\(p\)</span> is the number of markers. <span class="math inline">\(\mathbf{Y}\)</span> is a <span class="math inline">\((n \times q)\)</span> matrix of expression values for <span class="math inline">\(q\)</span> yeast genes. <span class="math inline">\(n\)</span> is again the numer of samples (112). <span class="math inline">\(\mathbf{GY}\)</span> is a <span class="math inline">\((q \times q)\)</span> adjacency matrix representing protein-protein interactions for the <span class="math inline">\(q\)</span> response variables.</p>
<p>We only use a smaller network in order to be able to print the results here.</p>
<div class="sourceCode" id="cb25"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb25-1" data-line-number="1">fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.Y=</span>G.Y, <span class="dt">lambda=</span><span class="dv">5</span>, <span class="dt">family=</span>gaussian, <span class="dt">maxit=</span><span class="dv">10</span>, <span class="dt">thresh=</span><span class="fl">1e-3</span>)</a>
<a class="sourceLine" id="cb25-2" data-line-number="2"><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/summary">summary</a></span>(fit)</a></code></pre></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.Y = G.Y, lambda = 5, thresh = 0.001, 
##     maxit = 10, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      5      0      1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>For the response variables we use an affinity matrix that represents <em>biological relationships</em> with <span class="math inline">\(\mathbf{GY}\)</span>. We promote sparsity by setting <span class="math inline">\(\lambda = 5\)</span> and put a weak (default) penalty on similar coefficients by setting <span class="math inline">\(\psi_{gy} = 1\)</span>. Other than that we used standard parameters in this case.</p>
<p>The <code>fit</code> object contains information about coefficients and intercepts. Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb27"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb27-1" data-line-number="1">X.new &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/matrix">matrix</a></span>(<span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/Normal">rnorm</a></span>(<span class="dv">10</span> <span class="op">*</span><span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/base/topics/nrow">ncol</a></span>(X)), <span class="dv">10</span>)</a>
<a class="sourceLine" id="cb27-2" data-line-number="2">pred  &lt;-<span class="st"> </span><span class="kw"><a href="https://www.rdocumentation.org/packages/stats/topics/predict">predict</a></span>(fit, X.new)</a>
<a class="sourceLine" id="cb27-3" data-line-number="3">pred[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>, <span class="dv">1</span><span class="op">:</span><span class="dv">5</span>]</a></code></pre></div>
<pre><code>##             [,1]       [,2]       [,3]       [,4]       [,5]
##  [1,]  4.1744776  4.1751662  4.1744079  4.1754934  4.1743196
##  [2,]  1.8105405  1.8122698  1.8111227  1.8135863  1.8113476
##  [3,]  1.8887596  1.8892567  1.8879708  1.8891502  1.8881457
##  [4,] -0.9665237 -0.9665011 -0.9677862 -0.9670347 -0.9673493
##  [5,]  6.2822988  6.2822908  6.2817507  6.2817011  6.2813731
##  [6,]  3.3963463  3.3961869  3.3966142  3.3960405  3.3965931
##  [7,]  5.3429670  5.3397246  5.3409361  5.3385110  5.3405420
##  [8,]  4.4231556  4.4209189  4.4189715  4.4203318  4.4195920
##  [9,]  2.9276149  2.9294643  2.9274470  2.9300377  2.9278901
## [10,]  1.4565552  1.4549309  1.4558023  1.4547471  1.4557673</code></pre>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-brem2005genetic">
<p>Brem, Rachel B, John D Storey, Jacqueline Whittle, and Leonid Kruglyak. 2005. “Genetic Interactions Between Polymorphisms That Affect Gene Expression in Yeast.” <em>Nature</em> 436 (7051). NIH Public Access: 701.</p>
</div>
<div id="ref-cheng2014graph">
<p>Cheng, Wei, Xiang Zhang, Zhishan Guo, Yu Shi, and Wei Wang. 2014. “Graph-Regularized Dual Lasso for Robust eQTL Mapping.” <em>Bioinformatics</em> 30 (12). Oxford Univ Press: i139–i148.</p>
</div>
<div id="ref-chung1997spectral">
<p>Chung, Fan RK, and Fan Chung Graham. 1997. <em>Spectral Graph Theory</em>. 92. American Mathematical Soc.</p>
</div>
<div id="ref-dirmeier2018netreg">
<p>Dirmeier, Simon, Christiane Fuchs, Nikola S Mueller, and Fabian J Theis. 2018. “netReg: Network-regularized linear models for biological association studies.” <em>Bioinformatics</em> 34 (5): 896–98.</p>
</div>
<div id="ref-powell2009bobyqa">
<p>Powell, Michael JD. 2009. “The Bobyqa Algorithm for Bound Constrained Optimization Without Derivatives.” <em>Cambridge NA Report NA2009/06, University of Cambridge, Cambridge</em>. Citeseer.</p>
</div>
<div id="ref-storey2005multiple">
<p>Storey, John D, Joshua M Akey, and Leonid Kruglyak. 2005. “Multiple Locus Linkage Analysis of Genomewide Expression in Yeast.” <em>PLoS Biology</em> 3 (8). Public Library of Science: e267.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#edgenet-tutorial">Edgenet tutorial</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Simon Dirmeier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.3.0.</p>
</div>
      </footer>
</div>

  

  </body>
</html>
