<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>netReg R tutorial • netReg</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/css/bootstrap.min.css" rel="stylesheet" integrity="sha384-BVYiiSIFeK1dGmJRAkycuHAHRg32OmUcww7on3RYdg4Va+PmSTsz/K68vbdEjh4u" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">netReg</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/netReg_R.html">R</a>
</li>
<li>
  <a href="../articles/netReg_commandline.html">C++</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/dirmeier/netReg">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>netReg R tutorial</h1>
            
          </div>

    
    
<div class="contents">
<p>This document explains how to install and use the <code>R</code> package.</p>
<div id="installation" class="section level2">
<h2 class="hasAnchor">
<a href="#installation" class="anchor"></a>Installation</h2>
<p>You can install and use <code>netReg</code> either</p>
<ul>
<li>as an <code>R</code> library from <a href="https://bioconductor.org/packages/release/bioc/html/netReg.html">Bioconductor</a>,</li>
<li>or by downloading the <code>tarball</code>.</li>
</ul>
<div id="installation-with-bioconductor" class="section level3">
<h3 class="hasAnchor">
<a href="#installation-with-bioconductor" class="anchor"></a>Installation with Bioconductor</h3>
<p>The easiest way to install <code>netReg</code> is using the <code>R</code>-console:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  <span class="kw">source</span>(<span class="st">"https://bioconductor.org/biocLite.R"</span>)
  <span class="kw">biocLite</span>(<span class="st">"netReg"</span>)
  
  <span class="kw">library</span>(netReg)</code></pre></div>
</div>
<div id="manual-installation" class="section level3">
<h3 class="hasAnchor">
<a href="#manual-installation" class="anchor"></a>Manual installation</h3>
<p>If you want to build the package yourself first download the <code>tarball</code> of the latest <a href="https://github.com/dirmeier/netReg/releases/">release</a> and then call:</p>
<div class="sourceCode"><pre class="sourceCode bash"><code class="sourceCode bash">  <span class="kw">R</span> CMD install <span class="kw">&lt;</span>netReg-x.y.z.tar.gz<span class="kw">&gt;</span></code></pre></div>
<p>That’s it.</p>
</div>
</div>
<div id="usage" class="section level2">
<h2 class="hasAnchor">
<a href="#usage" class="anchor"></a>Usage</h2>
<p>This section explains how to fit a linear model and do parameter estimation.</p>
<p>At first we load the library and some data:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  &gt;<span class="st"> </span><span class="kw">library</span>(netReg)
  &gt;<span class="st"> </span><span class="kw">data</span>(<span class="st">"yeast"</span>)
  
  &gt;<span class="st"> </span><span class="kw">ls</span>(yeast)
  &gt;<span class="st"> </span>[<span class="dv">1</span>] <span class="st">"GY"</span> <span class="st">"X"</span>  <span class="st">"Y"</span> </code></pre></div>
<p>The <code>yeast</code> data <code>X</code> and <code>Y</code> set is taken and adopted from <span class="citation">(Brem et al. 2005)</span>, <span class="citation">(Storey, Akey, and Kruglyak 2005)</span>, and <span class="citation">(Cheng et al. 2014)</span>. <span class="math inline">\(\mathbf{GY}\)</span> is taken from <a href="https://thebiogrid.org/downloads/archives/Release%20Archive/BIOGRID-3.4.150/BIOGRID-ORGANISM-3.4.150.tab.zip">BioGRID</a>. <code>X</code> is a <code>(n x p)</code> matrix of genetic markers where <code>n</code> is the number of samples (112) and <code>p</code> is the number of markers. The data has been taken from Brem <i>et al.</i>, Nature (2005). * <code>Y</code> is a <code>(n x q)</code> matrix of expression values for <code>q</code> yeast genes. <code>n</code> is again the numer of samples (112). * <code>GY</code> is a <code>(q x q)</code> adjacency matrix representing protein-protein interactions for the <code>q</code> response variables.</p>
<div id="model-fitting" class="section level3">
<h3 class="hasAnchor">
<a href="#model-fitting" class="anchor"></a>Model fitting</h3>
<p>Fitting a model using edge-based regularization with <code>netReg</code> is easy:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">  fit &lt;-<span class="st"> </span><span class="kw"><a href="../reference/edgenet.html">edgenet</a></span>(<span class="dt">X=</span>yeast$X, <span class="dt">Y=</span>yeast$Y, <span class="dt">G.Y=</span>yeast$GY,
                 <span class="dt">lambda=</span><span class="dv">10</span>, <span class="dt">psigy=</span><span class="dv">1</span>, <span class="dt">family=</span><span class="st">"gaussian"</span>)
  <span class="kw">print</span>(fit)</code></pre></div>
<p>For the response variables we use an affinity matrix that represents <em>biological relationships</em> with <code>GY</code>.</p>
<p>In this case we used a single affinity matrix <code>G.X</code> which represents the relationship of the covariables <span class="math inline">\(\mathit{X}\)</span>. If the design matrix has <span class="math inline">\(p\)</span> variables, <code>G.X</code> has to be an <span class="math inline">\((p \times p)\)</span>-dimensional symmetric matrix. We can also include a matrix for the response matrix <code>Y</code> with <span class="math inline">\(q\)</span> dependent variables. In that case the affinity matrix <code>G.Y</code> has to be <span class="math inline">\((q \times q)\)</span>-dimensional (and also symmetric).</p>
<p>The <code>fit</code> object contains information about coefficients, intercepts, residuals, etc. Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    X.new &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">10</span>*<span class="dv">5</span>),<span class="dv">10</span>)
    pred  &lt;-<span class="st"> </span><span class="kw">predict</span>(fit, X.new)</code></pre></div>
<p>The <code>pred</code> objects contains the predicted values for the responses.</p>
</div>
<div id="model-selection" class="section level3">
<h3 class="hasAnchor">
<a href="#model-selection" class="anchor"></a>Model selection</h3>
<p>In most cases we do not have the optimal shrinkage parameters <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_{gx}\)</span> and <span class="math inline">\(\psi_{gy}\)</span>. For these settings you can use <code>netReg</code>’s included model selection. We use Powell’s BOBYQA algorithm (<span class="citation">(Powell 2009)</span>) for gradient-free optimization that is included in the <code>C++</code> library <code>Dlib</code>. Doing the model selection only requires calling <code>cv.edgnet</code>:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    cv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cv.edgenet.html">cv.edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>aff.mat, <span class="dt">family=</span><span class="st">"gaussian"</span>, <span class="dt">maxit=</span><span class="dv">1000</span>)
    <span class="kw">print</span>(cv)</code></pre></div>
<p>You can use the fitted parameters for the normal <code>edgenet</code> function. In this scenario <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_{gx}\)</span> and <span class="math inline">\(\psi_{gy}\)</span> should be roughly <span class="math inline">\(0\)</span> for three reasons:</p>
<ul>
<li>we had enough data and a small number of covariables (<span class="math inline">\(n &gt; p\)</span>), so we can find the <em>BLUE</em> estimator,</li>
<li>we created <code>X</code> and <code>Y</code> independent of each other,</li>
<li>our prior graph <code>aff.mat</code> had little weight.</li>
</ul>
<p>Let’s do a scenario where we need to shrink some coefficients, i.e. <span class="math inline">\(n \ll p\)</span>. We choose a small <span class="math inline">\(p\)</span>, such that the computation does not take too long.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">    p &lt;-<span class="st"> </span><span class="dv">25</span>
    X &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">10</span>*p), <span class="dv">10</span>)
    Y &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rnorm</span>(<span class="dv">10</span>*p), <span class="dv">10</span>)
    aff.mat &lt;-<span class="st"> </span><span class="kw">matrix</span>(<span class="kw">rgamma</span>(p *<span class="st"> </span>p, <span class="dv">5</span>, <span class="dv">1</span>), p)
    aff.mat &lt;-<span class="st"> </span>(<span class="kw">t</span>(aff.mat) +<span class="st"> </span>aff.mat)
    <span class="kw">diag</span>(aff.mat) &lt;-<span class="st"> </span><span class="dv">0</span>
    cv &lt;-<span class="st"> </span><span class="kw"><a href="../reference/cv.edgenet.html">cv.edgenet</a></span>(<span class="dt">X=</span>X, <span class="dt">Y=</span>Y, <span class="dt">G.X=</span>aff.mat, <span class="dt">family=</span><span class="st">"gaussian"</span>, <span class="dt">maxit=</span><span class="dv">1000</span>)
    <span class="kw">print</span>(cv)</code></pre></div>
<p>In the above example <span class="math inline">\(\lambda\)</span> should have changed quite a bit, while <span class="math inline">\(\psi_{gy}\)</span> should still be <span class="math inline">\(0\)</span>. Since we generated <code>aff.mat</code> randomly <span class="math inline">\(\psi_{gx}\)</span> should be roughly (or exact) zero as well. This makes sense intuitively since we did not put any biological relationships into the affinity matrices.</p>
</div>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-brem2005genetic">
<p>Brem, Rachel B, John D Storey, Jacqueline Whittle, and Leonid Kruglyak. 2005. “Genetic Interactions Between Polymorphisms That Affect Gene Expression in Yeast.” <em>Nature</em> 436 (7051). NIH Public Access: 701.</p>
</div>
<div id="ref-cheng2014graph">
<p>Cheng, Wei, Xiang Zhang, Zhishan Guo, Yu Shi, and Wei Wang. 2014. “Graph-Regularized Dual Lasso for Robust EQTL Mapping.” <em>Bioinformatics</em> 30 (12). Oxford Univ Press: i139–i148.</p>
</div>
<div id="ref-powell2009bobyqa">
<p>Powell, Michael JD. 2009. “The Bobyqa Algorithm for Bound Constrained Optimization Without Derivatives.” <em>Cambridge NA Report NA2009/06, University of Cambridge, Cambridge</em>. Citeseer.</p>
</div>
<div id="ref-storey2005multiple">
<p>Storey, John D, Joshua M Akey, and Leonid Kruglyak. 2005. “Multiple Locus Linkage Analysis of Genomewide Expression in Yeast.” <em>PLoS Biology</em> 3 (8). Public Library of Science: e267.</p>
</div>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
        <div id="tocnav">
      <h2 class="hasAnchor">
<a href="#tocnav" class="anchor"></a>Contents</h2>
      <ul class="nav nav-pills nav-stacked">
<li><a href="#installation">Installation</a></li>
      <li><a href="#usage">Usage</a></li>
      <li><a href="#references">References</a></li>
      </ul>
</div>
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by Simon Dirmeier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
