<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Network-regularized regression models • netReg</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://cdnjs.cloudflare.com/ajax/libs/bootswatch/3.4.0/lumen/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><link href="../extra.css" rel="stylesheet">
<meta property="og:title" content="Network-regularized regression models">
<meta property="og:description" content="netReg">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    <div class="container template-article">
      <header><div class="navbar navbar-inverse navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">netReg</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">1.12.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Vignettes
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/netReg.html">netReg</a>
    </li>
  </ul>
</li>
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li>
  <a href="../news/index.html">News</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="http://www.simon-dirmeier.net">
    <span class="fa fa-home fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="https://github.com/dirmeier/netReg">
    <span class="fa fa-github fa-lg"></span>
     
  </a>
</li>
<li>
  <a href="http://twitter.com/simon_dirmeier">
    <span class="fa fa fa fa-twitter-square fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="netReg_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Network-regularized regression models</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/dirmeier/netReg/blob/master/vignettes/netReg.Rmd"><code>vignettes/netReg.Rmd</code></a></small>
      <div class="hidden name"><code>netReg.Rmd</code></div>

    </div>

    
    
<div id="introduction" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#introduction" class="anchor"></a>Introduction</h2>
<p>This vignette covers an introduction to network-regularized generalized linear regression models (GLMs). Network-regularization makes use of graphs to model interactions of covariables and responses in generalized linear regression models and penalizes the coefficients in GLMs using this information.</p>
</div>
<div id="edgenet" class="section level2">
<h2 class="hasAnchor">
<a href="#edgenet" class="anchor"></a>Edgenet</h2>
<p><code>edgenet</code> uses a <span class="math inline">\((p \times p)\)</span>-dimensional affinity matrix <span class="math inline">\(\mathbf{G}_X \in \mathbb{R}_+^{p \times p}\)</span> to model the interaction of <span class="math inline">\(p\)</span> covariables in <span class="math inline">\(\mathbf{X} \in \mathbb{R}^{n \times p}\)</span> and analogously a matrix <span class="math inline">\(\mathbf{G}_Y \in \mathbb{R}_+^{q \times q}\)</span> to model interactions of <span class="math inline">\(q\)</span> response variables in <span class="math inline">\(\mathbf{Y} \in \mathbb{R}^{n \times q}\)</span>. The affinity matrices are used for regularization of regression coefficients like this:</p>
<p><span class="math display">\[\begin{equation}
\small
\begin{split}
\hat{\mathbf{B}} = \arg \min_{\mathbf{B}} \,
 &amp; -\ell(\mathbf{B}) + \lambda ||\mathbf{B} ||_1  \\  &amp; +
              \frac{\psi_1}{2}  \sum_{i &lt; j}^p
               || \boldsymbol \beta_{i,*} - \boldsymbol \beta_{j,*} ||_2^2 (\mathbf{G}_X)_{i,j}    \\  &amp; +
 \frac{\psi_2}{2}  \sum_{i &lt; j}^q || \boldsymbol \beta_{*,i} - \boldsymbol \beta_{*,j} ||_2^2 (\mathbf{G}_Y)_{i,j} \\
 = \arg \min_{\mathbf{B}} \,
 &amp;  -\ell(\mathbf{B})  + \lambda ||\mathbf{B} ||_1  \\  &amp; +
 \psi_1 \text{tr}\left( \mathbf{B}^T \left( \mathbf{D}_{\mathbf{G}_X} - \mathbf{G}_X \right) \mathbf{B}  \right) \\  &amp; +
 \psi_2 \text{tr}\left( \mathbf{B} \left( \mathbf{D}_{\mathbf{G}_Y} - \mathbf{G}_Y  \right) \mathbf{B}^T  \right)
\end{split}
\label{equ:graphreg}
\end{equation}\]</span></p>
<p>Matrix <span class="math inline">\(\mathbf{B} \in \mathbb{R}^{p \times q}\)</span> is the matrix of regression coefficients to be estimated. Vectors <span class="math inline">\(\boldsymbol \beta_{i, *}\)</span> and <span class="math inline">\(\boldsymbol \beta_{*, i}\)</span> are the <span class="math inline">\(i\)</span>-th row or column of <span class="math inline">\(\mathbf{B}\)</span>, respectively. Shrinkage parameters <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_1\)</span> and <span class="math inline">\(\psi_2\)</span> are fixed or need to be estimated (e.g., using cross-validation). The matrices <span class="math inline">\(\mathbf{D}_{\mathbf{G}} - \mathbf{G}\)</span> are the combinatorial (or normalized) graph Laplacians of an affinity matrix <span class="math inline">\(\mathbf{G}\)</span> <span class="citation">(Chung and Graham 1997)</span>.</p>
</div>
<div id="group-lasso" class="section level2">
<h2 class="hasAnchor">
<a href="#group-lasso" class="anchor"></a>Group Lasso</h2>
<p>Using <code>group.lasso</code> estimates the coefficients as:</p>
<p><span class="math display">\[\begin{equation}
\small
\begin{split}
\hat{\mathbf{B}} = \arg \min_{\mathbf{B}}-\ell(\mathbf{B})  + \lambda q \sum_{g=1}^G \sqrt{p_{\mathcal{I}_g} } ||\mathbf{B}_{\mathcal{I}_g} ||_2
\end{split}
\end{equation}\]</span></p>
<p>where <span class="math inline">\(\mathcal{I}_g\)</span> is a set of indexes belonging to the <span class="math inline">\(g\)</span>th group of variables and <span class="math inline">\(p_{\mathcal{I}_g}\)</span> is the number of indexes in <span class="math inline">\(I_g\)</span>.</p>
<p>For details see <span class="citation">Yuan and Lin (2006)</span> and <span class="citation">Meier, Van De Geer, and Bühlmann (2008)</span>.</p>
</div>
<div id="families" class="section level2">
<h2 class="hasAnchor">
<a href="#families" class="anchor"></a>Families</h2>
<p>So far <code>netReg</code> supports the following exponential family random variables:</p>
<ul>
<li>Gaussian,</li>
<li>Binomial,</li>
<li>Poisson.</li>
</ul>
<p>The log-likelihood function <span class="math inline">\(\ell(\mathbf{B})\)</span> over <span class="math inline">\(q\)</span> response variables is defined as:</p>
<p><span class="math display">\[\begin{equation}
\small
\ell(\mathbf{B}) = \sum_{j}^q \sum_i^n \log \ \mathbb{P}\left({y}_{i, j} \mid h\left(\mathbf{x}_{i,*} \cdot \beta_{*,j}\right), \phi \right)
\end{equation}\]</span></p>
<p>where <span class="math inline">\(h\)</span> is the inverse of a link function, such as the logarithm, and <span class="math inline">\(\phi\)</span> is a dispersion parameter.</p>
</div>
<div id="fitting-a-model-to-data" class="section level2">
<h2 class="hasAnchor">
<a href="#fitting-a-model-to-data" class="anchor"></a>Fitting a model to data</h2>
<p>The following section shows how to use network regularization models. We shall use <code>edgenet</code>, but the calls for the other methods are analogous and examples can be found in the method documentation. We first load some libraries:</p>
<div class="sourceCode" id="cb1"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tensorflow</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">tfprobability</span>)
<span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">netReg</span>)</pre></body></html></div>
<pre><code>## 
## Attaching package: 'netReg'</code></pre>
<pre><code>## The following objects are masked from 'package:stats':
## 
##     binomial, family, gaussian, inverse.gaussian, poisson</code></pre>
<pre><code>## The following objects are masked from 'package:base':
## 
##     beta, gamma</code></pre>
<p>Set some parameters and create affinity matrices:</p>
<div class="sourceCode" id="cb5"><html><body><pre class="r"><span class="co"># parameters</span>
<span class="no">n</span> <span class="kw">&lt;-</span> <span class="fl">100</span>
<span class="no">p</span> <span class="kw">&lt;-</span> <span class="fl">10</span>
<span class="no">q</span> <span class="kw">&lt;-</span> <span class="fl">10</span>

<span class="co"># affinity matrices</span>
<span class="no">G.X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/rWishart.html">rWishart</a></span>(<span class="fl">1</span>, <span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">p</span>))[,,<span class="fl">1</span>])
<span class="no">G.Y</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/MathFun.html">abs</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/rWishart.html">rWishart</a></span>(<span class="fl">1</span>, <span class="fl">10</span>, <span class="fu"><a href="https://rdrr.io/r/base/diag.html">diag</a></span>(<span class="no">q</span>))[,,<span class="fl">1</span>])</pre></body></html></div>
<p>We created the affinity matrix absolutely random, although in practice a <em>real</em> (biological) observed affinity matrix should be used, because in the end the affinity matrix decides the shrinkage of the coefficients.</p>
<p>The actual fit is straightforward. We create Gaussian data first and then fit the model:</p>
<div class="sourceCode" id="cb6"><html><body><pre class="r"><span class="co"># data</span>
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">n</span> * <span class="no">p</span>), <span class="no">n</span>)
<span class="no">B</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">p</span> * <span class="no">q</span>), <span class="no">p</span>)
<span class="no">Y</span> <span class="kw">&lt;-</span> <span class="no">X</span> <span class="kw">%*%</span> <span class="no">B</span> + <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">n</span> * <span class="no">q</span>, <span class="fl">0</span>, <span class="fl">0.1</span>), <span class="no">n</span>)

<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">X</span>, <span class="kw">Y</span><span class="kw">=</span><span class="no">Y</span>, <span class="kw">G.X</span><span class="kw">=</span><span class="no">G.X</span>, <span class="kw">G.Y</span><span class="kw">=</span><span class="no">G.Y</span>, <span class="kw">family</span><span class="kw">=</span><span class="no">gaussian</span>, <span class="kw">maxit</span><span class="kw">=</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">fit</span>)</pre></body></html></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, maxit = 10, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      0      0      0 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>The <code>fit</code> object contains information about coefficients, intercepts etc.</p>
<div class="sourceCode" id="cb8"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span>(<span class="no">fit</span>)[,<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##                    y[1]         y[2]        y[3]        y[4]         y[5]
## (Intercept) -0.16114461 -0.231101796 -0.02932189  0.34179395 -0.105958976
## x[1]        -0.35954386 -0.231294900  0.16579993  0.34086064 -0.303890139
## x[2]        -0.39882138 -0.157718495  0.35752228 -0.18159449 -0.381166339
## x[3]        -0.18119393 -0.539452672 -0.32230726  0.13033013  0.141765386
## x[4]        -0.46252739 -0.005462316 -0.05451121 -0.06996223 -0.323262990
## x[5]        -0.22656828  0.012557382 -0.12988976 -0.12186553 -0.113724358
## x[6]        -0.04571471  0.443578422  0.25547269 -0.06740028  0.575974524
## x[7]        -0.26745677 -0.201941147  0.07049499 -0.25125182  0.137629658
## x[8]         0.16741446 -0.256720543 -0.18907519  0.06640834  0.001156025
## x[9]         0.43076041 -0.219526261 -0.08596466  0.29851440  0.179336891
## x[10]        0.28834793  0.347387195 -0.35431993 -0.05432715  0.344069093</code></pre>
<p>Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb10"><html><body><pre class="r"><span class="no">pred</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">fit</span>, <span class="no">X</span>)
<span class="no">pred</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##            [,1]        [,2]        [,3]        [,4]       [,5]
## [1,] -0.2870710 -0.74437915 -0.57742387 -0.05883890 -0.2489479
## [2,] -0.6241154  0.02684638 -1.10174492 -1.53995221 -1.1145034
## [3,] -0.3343998 -1.15399477 -0.25568816  0.65562922 -0.7593250
## [4,]  0.3293761 -0.71543876  0.04887515  0.49480959  1.5112120
## [5,] -0.3177100  0.02267839  0.10488057 -0.07972044  0.4552460</code></pre>
<p>The binomial case is the same only with a different <code>family</code> argument:</p>
<div class="sourceCode" id="cb12"><html><body><pre class="r"><span class="co"># data</span>
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">n</span> * <span class="no">p</span>), <span class="no">n</span>)
<span class="no">B</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">p</span> * <span class="no">q</span>), <span class="no">p</span>)

<span class="no">eta</span> <span class="kw">&lt;-</span> <span class="fl">1</span> / (<span class="fl">1</span> + <span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(-<span class="no">X</span> <span class="kw">%*%</span> <span class="no">B</span>))
<span class="no">Y.binom</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">10</span>), <span class="kw">function</span>(<span class="no">.</span>) <span class="fu"><a href="https://rdrr.io/r/stats/Binomial.html">rbinom</a></span>(<span class="no">n</span>, <span class="fl">1</span>, <span class="no">eta</span>[,<span class="no">.</span>])))

<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">X</span>, <span class="kw">Y</span><span class="kw">=</span><span class="no">Y</span>, <span class="kw">G.X</span><span class="kw">=</span><span class="no">G.X</span>, <span class="kw">G.Y</span><span class="kw">=</span><span class="no">G.Y</span>, <span class="kw">family</span><span class="kw">=</span><span class="no">binomial</span>, <span class="kw">maxit</span><span class="kw">=</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">fit</span>)</pre></body></html></div>
<pre><code>## 'binomial.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, maxit = 10, family = binomial)
## 
## parameters:
## lambda  psigx  psigy 
##      0      0      0 
## 
## family:  binomial
## link:  logit 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>The Poisson case of course works analogously:</p>
<div class="sourceCode" id="cb14"><html><body><pre class="r"><span class="co"># data</span>
<span class="no">X</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">n</span> * <span class="no">p</span>), <span class="no">n</span>)
<span class="no">B</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="no">p</span> * <span class="no">q</span>), <span class="no">p</span>)

<span class="no">eta</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/Log.html">exp</a></span>(-<span class="no">X</span> <span class="kw">%*%</span> <span class="no">B</span>)
<span class="no">Y.pois</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/do.call.html">do.call</a></span>(<span class="st">"cbind"</span>, <span class="fu"><a href="https://rdrr.io/r/base/lapply.html">lapply</a></span>(<span class="fu"><a href="https://rdrr.io/r/base/seq.html">seq</a></span>(<span class="fl">10</span>), <span class="kw">function</span>(<span class="no">.</span>) <span class="fu"><a href="https://rdrr.io/r/stats/Poisson.html">rpois</a></span>(<span class="no">n</span>, <span class="no">eta</span>[,<span class="no">.</span>])))

<span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">X</span>, <span class="kw">Y</span><span class="kw">=</span><span class="no">Y.pois</span>, <span class="kw">G.X</span><span class="kw">=</span><span class="no">G.X</span>, <span class="kw">G.Y</span><span class="kw">=</span><span class="no">G.Y</span>, <span class="kw">family</span><span class="kw">=</span><span class="no">poisson</span>, <span class="kw">maxit</span><span class="kw">=</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">fit</span>)</pre></body></html></div>
<pre><code>## 'poisson.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y.pois, G.X = G.X, G.Y = G.Y, maxit = 10, 
##     family = poisson)
## 
## parameters:
## lambda  psigx  psigy 
##      0      0      0 
## 
## family:  poisson
## link:  log 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
</div>
<div id="model-selection" class="section level2">
<h2 class="hasAnchor">
<a href="#model-selection" class="anchor"></a>Model selection</h2>
<p>In most cases we do not have the optimal shrinkage parameters <span class="math inline">\(\lambda\)</span>, <span class="math inline">\(\psi_{1}\)</span> and <span class="math inline">\(\psi_{2}\)</span>. For these settings you can use <code>netReg</code>’s included model selection. We use Powell’s BOBYQA algorithm <span class="citation">(Powell 2009)</span> for gradient-free optimization in a coss-validation framework. Doing the model selection only requires calling <code>cv.edgenet</code>:</p>
<div class="sourceCode" id="cb16"><html><body><pre class="r"><span class="no">cv</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/cvedgenet-methods.html">cv.edgenet</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">X</span>, <span class="kw">Y</span><span class="kw">=</span><span class="no">Y</span>, <span class="kw">G.X</span><span class="kw">=</span><span class="no">G.Y</span>, <span class="no">G.Y</span>, <span class="kw">family</span><span class="kw">=</span><span class="no">gaussian</span>, <span class="kw">optim.maxit</span><span class="kw">=</span><span class="fl">10</span>, <span class="kw">maxit</span><span class="kw">=</span><span class="fl">10</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">cv</span>)</pre></body></html></div>
<pre><code>## 'gaussian.cv.edgenet' object
## 
## call:
## cv.edgenet(X = X, Y = Y, G.X = G.Y, G.Y = G.Y, maxit = 10, family = gaussian, 
##     optim.maxit = 10)
## 
## optimal parameters:
## lambda (estimated)  psigx (estimated)  psigy (estimated) 
##                0.0                0.0                0.1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(cv) for coefficients</code></pre>
<p>The cross-validation and BOBYQA is quite computationally intensive, so we set <code>optim.maxit</code> to <span class="math inline">\(10\)</span>. I recommend using <code>TensorFlow</code> on a GPU for that, since <code>edgenet</code> needs to compute many matrix multiplications.</p>
<p>The <code>cv.edgenet</code> object also contains a fit with the optimal parameters.</p>
<div class="sourceCode" id="cb18"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">cv</span>$<span class="no">fit</span>)</pre></body></html></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.X = G.X, G.Y = G.Y, lambda = ret$lambda, 
##     psigx = ret$psigx, psigy = ret$psigy, thresh = thresh, maxit = maxit, 
##     learning.rate = learning.rate, family = family)
## 
## parameters:
## lambda  psigx  psigy 
##    0.0    0.0    0.1 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(cv$fit) for coefficients</code></pre>
<p>You can obtain the coefficients like this:</p>
<div class="sourceCode" id="cb20"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/stats/coef.html">coef</a></span>(<span class="no">cv</span>)[,<span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##                    y[1]        y[2]        y[3]        y[4]         y[5]
## (Intercept) -0.16077435 -0.23127170 -0.02970739  0.34189507 -0.111753576
## x[1]        -0.16237111 -0.23148572 -0.02958075  0.34068984 -0.105691418
## x[2]        -0.39972085 -0.34606296  0.16797587 -0.16731170 -0.381302536
## x[3]        -0.18187171 -0.34044415 -0.32307500  0.13098395  0.143781543
## x[4]        -0.46328911 -0.01993662 -0.05692720 -0.11718661 -0.124845296
## x[5]        -0.03006735  0.18756928 -0.29005367 -0.28308502 -0.117987476
## x[6]        -0.04212106  0.24586141  0.06996688  0.13110411  0.376966745
## x[7]        -0.07066823 -0.20133366  0.06610991 -0.25744832  0.141139224
## x[8]        -0.01823750 -0.25720704 -0.19019511 -0.13266508  0.003752189
## x[9]         0.23437476 -0.10024169  0.06756929  0.29905838  0.074085966
## x[10]        0.28875348  0.34813097 -0.35467392  0.01685912  0.344792455</code></pre>
<p>Furthermore, you can also directly predict using a <code>cv.edgenet</code> object:</p>
<div class="sourceCode" id="cb22"><html><body><pre class="r"><span class="no">pred</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">cv</span>, <span class="no">X</span>)
<span class="no">pred</span>[<span class="fl">1</span>:<span class="fl">5</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##            [,1]         [,2]       [,3]       [,4]        [,5]
## [1,]  0.7557590  0.346999779 -0.7708096  0.8892882  0.37526357
## [2,] -0.7116197 -1.005773856 -0.1176944 -0.8742186 -0.05700795
## [3,]  0.6350628 -0.986363955 -0.1685407  1.6333573  0.46848193
## [4,] -0.3026532 -0.005457719 -0.2907739  0.1246400 -0.32110764
## [5,] -0.1342012 -0.216802457 -0.2822949  0.4363335  0.18042592</code></pre>
</div>
<div id="a-biological-example" class="section level2">
<h2 class="hasAnchor">
<a href="#a-biological-example" class="anchor"></a>A biological example</h2>
<p>This section explains how to fit a linear model and do parameter estimation.</p>
<p>At first we load the library and some data:</p>
<div class="sourceCode" id="cb24"><html><body><pre class="r"><span class="fu"><a href="https://rdrr.io/r/base/library.html">library</a></span>(<span class="no">netReg</span>)
<span class="fu"><a href="https://rdrr.io/r/utils/data.html">data</a></span>(<span class="st">"yeast"</span>)

<span class="fu"><a href="https://rdrr.io/r/base/ls.html">ls</a></span>(<span class="no">yeast</span>)

<span class="no">X</span> <span class="kw">&lt;-</span> <span class="no">yeast</span>$<span class="no">X</span>
<span class="no">Y</span> <span class="kw">&lt;-</span> <span class="no">yeast</span>$<span class="no">Y</span>
<span class="no">G.Y</span> <span class="kw">&lt;-</span> <span class="no">yeast</span>$<span class="no">GY</span></pre></body></html></div>
<p>The <em>yeast</em> data <span class="math inline">\(\mathbf{X}\)</span> and <span class="math inline">\(\mathbf{Y}\)</span> set is taken and adopted from <span class="citation">(Brem et al. 2005)</span>, <span class="citation">(Storey, Akey, and Kruglyak 2005)</span>, and <span class="citation">(Cheng et al. 2014)</span>. <span class="math inline">\(\mathbf{GY}\)</span> is taken from <a href="https://thebiogrid.org/downloads/archives/Release%20Archive/BIOGRID-3.4.150/BIOGRID-ORGANISM-3.4.150.tab.zip">BioGRID</a>.</p>
<p><span class="math inline">\(\mathbf{X}\)</span> is a <span class="math inline">\((n \times p)\)</span> matrix of genetic markers where <span class="math inline">\(n\)</span> is the number of samples (112) and <span class="math inline">\(p\)</span> is the number of markers. <span class="math inline">\(\mathbf{Y}\)</span> is a <span class="math inline">\((n \times q)\)</span> matrix of expression values for <span class="math inline">\(q\)</span> yeast genes. <span class="math inline">\(n\)</span> is again the numer of samples (112). <span class="math inline">\(\mathbf{GY}\)</span> is a <span class="math inline">\((q \times q)\)</span> adjacency matrix representing protein-protein interactions for the <span class="math inline">\(q\)</span> response variables.</p>
<p>We only use a smaller network in order to be able to print the results here.</p>
<div class="sourceCode" id="cb25"><html><body><pre class="r"><span class="no">fit</span> <span class="kw">&lt;-</span> <span class="fu"><a href="../reference/edgenet-methods.html">edgenet</a></span>(<span class="kw">X</span><span class="kw">=</span><span class="no">X</span>, <span class="kw">Y</span><span class="kw">=</span><span class="no">Y</span>, <span class="kw">G.Y</span><span class="kw">=</span><span class="no">G.Y</span>, <span class="kw">lambda</span><span class="kw">=</span><span class="fl">5</span>, <span class="kw">family</span><span class="kw">=</span><span class="no">gaussian</span>, <span class="kw">maxit</span><span class="kw">=</span><span class="fl">10</span>, <span class="kw">thresh</span><span class="kw">=</span><span class="fl">1e-3</span>)
<span class="fu"><a href="https://rdrr.io/r/base/summary.html">summary</a></span>(<span class="no">fit</span>)</pre></body></html></div>
<pre><code>## 'gaussian.edgenet' object
## 
## call:
## edgenet(X = X, Y = Y, G.Y = G.Y, lambda = 5, thresh = 0.001, 
##     maxit = 10, family = gaussian)
## 
## parameters:
## lambda  psigx  psigy 
##      5      0      0 
## 
## family:  gaussian
## link:  identity 
## 
## -&gt; call coef(fit) for coefficients</code></pre>
<p>For the response variables we use an affinity matrix that represents <em>biological relationships</em> with <span class="math inline">\(\mathbf{GY}\)</span>. We promote sparsity by setting <span class="math inline">\(\lambda = 5\)</span> and put a weak (default) penalty on similar coefficients by setting <span class="math inline">\(\psi_{gy} = 1\)</span>. Other than that we used standard parameters in this case.</p>
<p>The <code>fit</code> object contains information about coefficients and intercepts. Having the coefficients estimated we are able to predict novel data-sets:</p>
<div class="sourceCode" id="cb27"><html><body><pre class="r"><span class="no">X.new</span> <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html">matrix</a></span>(<span class="fu"><a href="https://rdrr.io/r/stats/Normal.html">rnorm</a></span>(<span class="fl">10</span> * <span class="fu"><a href="https://rdrr.io/r/base/nrow.html">ncol</a></span>(<span class="no">X</span>)), <span class="fl">10</span>)
<span class="no">pred</span>  <span class="kw">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/stats/predict.html">predict</a></span>(<span class="no">fit</span>, <span class="no">X.new</span>)
<span class="no">pred</span>[<span class="fl">1</span>:<span class="fl">10</span>, <span class="fl">1</span>:<span class="fl">5</span>]</pre></body></html></div>
<pre><code>##             [,1]       [,2]       [,3]       [,4]        [,5]
##  [1,] -0.8044893 -0.4087922 -0.4094400  0.7983694  0.19921542
##  [2,] -2.2041351 -1.2107032 -0.9572445 -0.8381094 -0.47034967
##  [3,]  0.3849456 -0.4734009  0.1799251  0.2595439 -0.48178111
##  [4,] -0.3617251  0.7427117  0.7250823  0.3922634 -0.05791476
##  [5,] -0.2753576 -0.7413358 -0.9715962  0.3449987  0.13214298
##  [6,] -0.6166750 -0.9227181 -1.2528798  0.1904363  0.19282596
##  [7,]  1.4740172  0.5679683  0.5016382 -0.1711926  0.61066369
##  [8,] -0.1751318 -0.4029924  0.7903120  0.3881409 -0.89040746
##  [9,]  1.1178270  0.7140209 -0.1323093  0.3601851  0.40989170
## [10,] -0.3520390 -1.3073201 -0.6078369  0.4341101 -0.15034964</code></pre>
</div>
<div id="references" class="section level2 unnumbered">
<h2 class="hasAnchor">
<a href="#references" class="anchor"></a>References</h2>
<div id="refs" class="references">
<div id="ref-brem2005genetic">
<p>Brem, Rachel B, John D Storey, Jacqueline Whittle, and Leonid Kruglyak. 2005. “Genetic Interactions Between Polymorphisms That Affect Gene Expression in Yeast.” <em>Nature</em> 436 (7051): 701.</p>
</div>
<div id="ref-cheng2014graph">
<p>Cheng, Wei, Xiang Zhang, Zhishan Guo, Yu Shi, and Wei Wang. 2014. “Graph-Regularized Dual Lasso for Robust eQTL Mapping.” <em>Bioinformatics</em> 30 (12): i139–i148.</p>
</div>
<div id="ref-chung1997spectral">
<p>Chung, Fan RK, and Fan Chung Graham. 1997. <em>Spectral Graph Theory</em>. 92. American Mathematical Soc.</p>
</div>
<div id="ref-meier2008group">
<p>Meier, Lukas, Sara Van De Geer, and Peter Bühlmann. 2008. “The Group Lasso for Logistic Regression.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 70 (1): 53–71.</p>
</div>
<div id="ref-powell2009bobyqa">
<p>Powell, Michael JD. 2009. “The Bobyqa Algorithm for Bound Constrained Optimization Without Derivatives.” <em>Cambridge NA Report NA2009/06, University of Cambridge, Cambridge</em>.</p>
</div>
<div id="ref-storey2005multiple">
<p>Storey, John D, Joshua M Akey, and Leonid Kruglyak. 2005. “Multiple Locus Linkage Analysis of Genomewide Expression in Yeast.” <em>PLoS Biology</em> 3 (8): e267.</p>
</div>
<div id="ref-yuan2006model">
<p>Yuan, Ming, and Yi Lin. 2006. “Model Selection and Estimation in Regression with Grouped Variables.” <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 68 (1): 49–67.</p>
</div>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p>Developed by Simon Dirmeier.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="https://pkgdown.r-lib.org/">pkgdown</a> 1.5.1.</p>
</div>

      </footer>
</div>

  


  </body>
</html>
